src/jamsocket.ts

```diff
@@ -0,0 +1,88 @@
+import { JamsocketApi, ServiceCreateResult, ServiceListResult, SpawnRequestBody, SpawnResult, TokenCreateResult, TokenRequestBody, TokenRevokeResult } from './api'
+import { JamsocketConfig, readJamsocketConfig } from './common'
+import { ContainerManager, detectContainerManager } from './container-manager'
+
+export class Jamsocket {
+  constructor(private config: JamsocketConfig | null, private api: JamsocketApi) {}
+
+  public static fromEnvironment(): Jamsocket {
+    const config = readJamsocketConfig()
+    const api = JamsocketApi.fromEnvironment()
+
+    return new Jamsocket(config, api)
+  }
+
+  public expectAuthorized(): JamsocketConfig {
+    if (this.config === null) {
+      throw new Error('No user credentials found. Log in with jamsocket login')
+    }
+
+    return this.config
+  }
+
+  public async serviceImage(service: string): Promise<string> {
+    const config = this.expectAuthorized()
+    const result = await this.api.serviceImage(config.username, service, config.auth)
+    return result.imageName
+  }
+
+  public async push(service: string, image: string, tag?: string): Promise<void> {
+    const config = this.expectAuthorized()
+    const containerManager: ContainerManager = detectContainerManager()
+
+    let prefixedImage = await this.serviceImage(service)
+    if (tag) prefixedImage += `:${tag}`
+
+    console.log('Tagging.')
+    containerManager.tag(image, prefixedImage)
+
+    console.log('Pushing.')
+    await containerManager.push(prefixedImage, config.auth)
+
+    console.log('Done.')
+  }
+
+  public spawn(service: string, env: Record<string, string>, grace?: number, port?: number, tag?: string): Promise<SpawnResult> {
```

Comment:
I think `env` should be optional here. Not sure why it's not throwing a typecheck error in the `spawn` command though 🤔

Reply:
Fixed.